### DISCUSSION TITLE
IPSec tunnels fail on bad mix of Initiators/Responders

### DISCUSSION THREAD

**Discussion by martincco** [strongswan/strongswan#1023] [Q&A]
I am seeing a problem when I create IPSec tunnels in a circular network, when each device is the Initiator on one tunnel and a Responder on the other tunnel.  Maybe a diagram will help show what I mean.

![image](https://user-images.githubusercontent.com/104164388/165837144-73e89505-36dd-4254-8e6e-70fce8ded210.png)

I am using strongswan 5.8.4.
Each device has a Tunnel to each of the other 2 devices.  And each device is an Initiator on one of its tunnels and the Responder on the other. When I do this, 2 of the tunnels get created and I can send data through them. But the other tunnel does not get created successfully.  After 3 minutes a different tunnel goes down, and the one that was down comes up.  For example:

•	Create Tunnel A.  Wait 5 minutes
•	Create Tunnel B.  Wait 5 minutes
•	Create Tunnel C.  Wait 5 minutes
•	Tunnels B and C come up properly, and data flows through them. Tunnel A does not come up
•	After 3 minutes, all tunnels go down briefly for a new attempt. Only tunnels C and A come up
•	After 3 more minutes, all tunnels go down briefly, only tunnels A and B come up
•	After 3 more minutes, all tunnels go down briefly, only tunnels B and C come up
•	This cycle repeats every approximately three minutes forever.

If I do not wait between creating the tunnels, and create them at roughly the same time, I sometimes see this behavior, but not always.  If I swap the Initiator and Responder on one of the tunnels, (so that, for example, Dev 1 is the Initiator on both of its tunnels, Dev 2 is the Initiator on one of its tunnels, and Dev 3 is the Responder on both of its tunnels) all 3 tunnels come up right away and we do not have this problem at all. It is only when each device is both an Initiator and a Responder.  It only occurs when we have this circular network.

I have tried this using all IKEv2 password-based tunnels. I have tried it with all IKEv1 password-based tunnels. I have tried it with all IKEv2 Certificate-based tunnels. And I have tried it using a mix of all 3 of those profiles.  I got the same results in each case.

I have added in a 4th device and connected Dev 3 to Dev 4 and Dev 4 to Dev 1 with tunnels (removing the tunnel from Dev 3 to Dev 1). I saw the same behavior.  If I have just 2 devices, with 2 tunnels between them, and set each DUT as an Initiator and Responder, I also see the behavior, though not every time.

In charon.conf (attached in the zip file), the only value not default is max_ikev1_exchanges.  I bumped that to 10 as I was having trouble connecting some of my IKEv1 tunnels.

I have tried massaging the following charon.conf parameters, but none seems to correct this behavior.
block_threshold
half_open_timeout
ikesa_limit
init_limit_half_open
max_ikev1_exchanges  (I nominally have this at 10 to correct other issues, I tried bumping it higher)
retransmit_base
retransmit_timeout
retransmit_tries
threads
host_resolver.max_threads

In ipsec.conf file (in the attached zip file) each device uses the same file. I think the main thing of interest there is that
     keyingtries=%forever

I have attached logs for all 3 devices (in the zip file), covering starting up ipsec on all devices, and then 2 full cycles of rotating tunnels going up and down.  Here is a timeline of the main events, so you can quickly find them in the logs:
•	14:01	Start ipsec on device 1
•	14:06	Start ipsec on device 2
•	14:07	Tunnel A comes up and data flows successfully
•	14:11	Start ipsec on device 3
•	14:12	Tunnel A goes down.  Tunnels B & C come up and data flows successfully
•	14:14	Tunnel C goes down.  Tunnel A comes up. Data flows on Tunnels A & B
•	14:17	Tunnel B goes down.  Tunnel C comes up. Data flows on Tunnels C & A
•	14:20	Tunnel A goes down.  Tunnel B comes up. Data flows on Tunnels B &C
•	14:23	Tunnel C goes down.  Tunnel A comes up. Data flows on Tunnels A & B
•	14:27	Tunnel B goes down.  Tunnel C comes up. Data flows on Tunnels C & A
•	14:30	Tunnel A goes down.  Tunnel B comes up. Data flows on Tunnels B & C
•	14:31	Stop ipsec on device 3
•	14:31	Stop ipsec on device 2
•	14:32	Stop ipsec on device 1

I would appreciate any suggestions you might have for keeping all 3 tunnels up and data flowing across all 3 tunnels.  

I have attached ipsec statusall for all 3 devices.  It was collected when Tunnel A and Tunnel B were up, and Tunnel C was down
[strongswan_files.zip](https://github.com/strongswan/strongswan/files/8586042/strongswan_files.zip)
.

**Comment by tobiasbrunner** [comment]
What I do see in the logs are restarts of the IKE daemon around some of the timestamps you indicate, which definitely will cause existing tunnels to go down (on dev1 at 14:13 and 14:23, on dev2 at 14:07, 14:10, 14:20, and 14:29, and on dev3 at 14:17 and 14:26). So who or what does that exactly (and why)?

**Reply by martincco** [reply]
Sorry it took so long to get back.  I wanted to make sure I dug into this to get some better answers, and it turns out I was able to simplify it significantly. I used the "triangle" of devices above, because that is how I first ran into the problem. But it turns out I can reproduce the problem with a single tunnel.  In addition, I discovered that I have an "ipsec watchdog" running that restarts ipsec every 3 minutes if expected tunnels are not found.  That is what was causing the stops/starts you found.  Again, I apologize for not understanding my context better.

So, here is what I am seeing now.  If I have a single tunnel, the tunnel comes up just fine and data is flowing.  If I then stop and then start ipsec on the Responder side, the tunnel stays down and never comes back up again.  I have attached new logs, and other files. ipsec.conf and charon.conf have not changed since my first post. The only changes to ip addr and ip route are that I removed the other two tunnels.

Here is the timeline of events now:

8:55:00 - ipsec start on dev 1 (Initiator)
8:56:00 - ipsec start on dev 2 (Responder)
8:56:39 - tunnel comes up, data flows
8:59:00 - command ipsec stop on dev 2, tunnel goes down, no data flowing
8:59:10 - command ipsec start on dev 2
9:04:00 - tunnel still down, still no data.  This is when I grabbed 'ipsec statusall' on both devices
9:04 - 9:06 - collect all log and config files
9:07:00 - ipsec stop on both dev 1 and dev 2

So, after stopping ipsec at 8:59 and bringing it back up (all on the Responder side), the tunnel never comes back up.  I have tested this with basically 0 seconds between stop and start.  The timeline above shows 10 seconds. And I have also tested up to several minutes.  In every case, the tunnel never comes back up.  I think the Initiator just does not realize that the tunnel went down.  The only way to get the tunnel to come back up is the stop and start ipsec on the Initiator side.
Thanks for the help
[strongswanfiles_b.zip](https://github.com/strongswan/strongswan/files/8611815/strongswanfiles_b.zip)
.

**Reply by tobiasbrunner** [reply]
> I discovered that I have an "ipsec watchdog" running that restarts ipsec every 3 minutes if expected tunnels are not found

Never heard of this. Where is that from? Did you write that yourself? (And restarting the whole daemon if a tunnel is not found doesn't seem to make sense to be honest. Wouldn't just initiating the missing tunnels be a better approach? Or using trap policies so they are (re-)started automatically if traffic matches.)

>  If I then stop and then start ipsec on the Responder side, the tunnel stays down and never comes back up again. 

No, because you haven't configured `closeaction=restart`, which would trigger a restart if the peer terminates the SA (read the [documentation](https://wiki.strongswan.org/projects/strongswan/wiki/Connsection) for some caveats), or `auto=route`, which would install trap policies so the connection is automatically restarted if traffic matches `left|rightsubnet` (in this case configuring `closeaction`, or `dpdaction=restart` is not recommended in versions before 5.9.6 as that could cause duplicate CHILD_SAs). It obviously also depends on what the other peer has configured (i.e. whether it is configured as responder only, or if it also automatically starts connections or uses trap policies itself).

**Reply by martincco** [reply]
> > I discovered that I have an "ipsec watchdog" running that restarts ipsec every 3 minutes if expected tunnels are not found
> 
> Never heard of this. Where is that from? Did you write that yourself? (And restarting the whole daemon if a tunnel is not found doesn't seem to make sense to be honest. Wouldn't just initiating the missing tunnels be a better approach? Or using trap policies so they are (re-)started automatically if traffic matches.)
> 

Yes, the watchdog is something that is in our code that was written a long time ago.  It is definitely not optimum and should be fixed so that on a dead tunnel, only that tunnel is re-initiated. 

> > If I then stop and then start ipsec on the Responder side, the tunnel stays down and never comes back up again.
> 
> No, because you haven't configured `closeaction=restart`, which would trigger a restart if the peer terminates the SA (read the [documentation](https://wiki.strongswan.org/projects/strongswan/wiki/Connsection) for some caveats), or `auto=route`, which would install trap policies so the connection is automatically restarted if traffic matches `left|rightsubnet` (in this case configuring `closeaction`, or `dpdaction=restart` is not recommended in versions before 5.9.6 as that could cause duplicate CHILD_SAs). It obviously also depends on what the other peer has configured (i.e. whether it is configured as responder only, or if it also automatically starts connections or uses trap policies itself).

Ah, this is what I was looking for. I read through all the docs for ipsec.conf and charon.conf, but didn’t notice the ‘route’ option for ‘auto’. I have now tested it, and it generally works as you described.  If the Responder side shuts down a tunnel and then restarts it, then tunnel will get reinitialized if traffic is sent from the Responder side. Sometimes the tunnel came back up just from traffic from the Initiator side, but not always.  I was not able to determine if there was a specific set of timing or order of steps that could allow this. It is likely a very tight timing issue, and you must send data from the Initiator at the right time while the Responder is starting up the tunnel.  That is my guess, could be totally wrong.

Thank you for your help. Between this config parameter and updating the watchdog we have, I can keep my tunnels up and recover properly if one gets reset.
